src/metabase/driver/h2.clj

```diff
@@ -127,22 +132,50 @@
       (when (instance? SessionLocal session)
         (Parser. session)))))
 
-(defn- contains-ddl?
+(defn- classify-query
+  "Takes an h2 db id, and a query, returns the command-classes+types (see: ), and any remaining sql.
+
+  - `remaining-sql` is a nillable sql string that is unable to be classified without running preceding queries first.
+    Usually if it exists we will deny the query."
   [database query]
   (when-let [h2-parser (make-h2-parser database)]
     (try
-      (let [command      (.prepareCommand h2-parser query)
-            command-type (.getCommandType command)]
-        ;; TODO: do we need to handle CommandList?
-        ;; Command types are organized with all DDL commands listed first
-        ;; see https://github.com/h2database/h2database/blob/master/h2/src/main/org/h2/command/CommandInterface.java
-        (< command-type CommandInterface/ALTER_SEQUENCE))
+      (let [command-list      (.prepareCommand h2-parser query)
+            this-command      ((juxt class #(.getCommandType ^org.h2.command.CommandList %)) command-list)
+
+            ;; when there are no fields: return no commands
+            prepared-commands (map (juxt class #(.getType ^org.h2.command.Prepared %))
+                                   (get-field command-list "commands" []))
+
+            ;; when there is no remaining sql: return nil for remaining-sql
+            remaining-sql (get-field command-list "remaining" nil)]
+        {:command-class+types (vec (concat [this-command] prepared-commands))
+         :remaining-sql       remaining-sql})
       ;; if the query is invalid, then it isn't DDL
-      (catch Throwable _ false))))
+      (catch Throwable e
+        ;; TODO: Put more stuff in here -- check that it threw v.s. detected no ddl.
+        (log/warn e)
+        e))))
+
+(defn- cmd-type-ddl? [cmd-type]
+  ;; Command types are organized with all DDL commands listed first, so all ddl commands are before ALTER_SEQUENCE.
+  ;; see https://github.com/h2database/h2database/blob/master/h2/src/main/org/h2/command/CommandInterface.java#L297
+  (< cmd-type CommandInterface/ALTER_SEQUENCE))
+
+(defn- contains-ddl? [{:keys [command-class+types remaining-sql]}]
+  (let [cmd-type-nums (mapv second command-class+types)]
```

Comment:
second value in each `class+type` here will correspond to a value in https://github.com/h2database/h2database/blob/master/h2/src/main/org/h2/command/CommandInterface.java#L16-L557.